noir-r1cs/src/sparse_matrix.rs

```diff
@@ -175,10 +182,28 @@ impl Mul<HydratedSparseMatrix<'_>> for &[FieldElement] {
             rhs.matrix.num_rows,
             "Vector length does not match number of rows."
         );
+
+        let num_threads = rayon::current_num_threads();
         let mut result = vec![FieldElement::zero(); rhs.matrix.num_cols];
-        for ((i, j), value) in rhs.iter() {
-            result[j] += value * self[i];
-        }
+
+        let chunk_size = result.len().div_ceil(num_threads);
+
+        // In microbenchmarks par_iter_mut.chunks outperforms par_chunks_mut slightly.
+        result
+            .par_iter_mut()
+            .chunks(chunk_size)
+            .enumerate()
+            .for_each(|(chunk_number, mut chunk)| {
+                let base = chunk_number * chunk_size;
+                let col_range = base..base + chunk_size;
```

Comment:
`let col_range = base..base + chunk.len()`. 

Otherwise `col_range` will be out of bounds when `chunk_size` doesn't divide `result.len()`. You are protected here from this bug because `col` will only be in-range, but better to do it right.

Reply:
`let base = chunk_number * chunk_size` should still be correct as only the last chunk is not exactly `chunk_size` (but please confirm with rayon docs, and maybe leave a comment explaining correctness.)